### PR TITLE
fixes #82: support classes

### LINKED ISSUE #82: ES6 Support: Classes
# [14.5](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-class-definitions) Class Definitions

## Syntax

<pre>
<i>ClassDeclaration</i>[Yield, Default] :
    <b>class</b> <i>BindingIdentifier</i><sub>[?Yield]</sub> <i>ClassTail</i><sub>[?Yield]</sub>
    [+Default] <b>class</b> <i>ClassTail</i><sub>[?Yield]</sub>

<i>ClassExpression</i>[Yield,GeneratorParameter] :
    <b>class</b> <i>BindingIdentifier</i><sub>[?Yield]</sub><sub>opt</sub> <i>ClassTail</i><sub>[?Yield,?GeneratorParameter]</sub>

<i>ClassTail</i>[Yield,GeneratorParameter] :
    [~GeneratorParameter] <i>ClassHeritage</i><sub>[?Yield]</sub><sub>opt</sub> <b>{</b> <i>ClassBody</i><sub>[?Yield]</sub><sub>opt</sub> <b>}</b>
    [+GeneratorParameter] <i>ClassHeritage<i><sub>opt</sub> <b>{</b> <i>ClassBody</i><sub>opt</sub></sub> <b>}</b>

<i>ClassHeritage</i>[Yield] :
    <b>extends</b> <i>LeftHandSideExpression</i><sub>[?Yield]</sub>

<i>ClassBody</i>[Yield] :
    <i>ClassElementList</i><sub>[?Yield]</sub>

<i>ClassElementList</i>[Yield] :
    <i>ClassElement</i><sub>[?Yield]</sub>
    <i>ClassElementList</i><sub>[?Yield]</sub> <i>ClassElement</i><sub>[?Yield]</sub>

<i>ClassElement</i>[Yield] :
    <i>MethodDefinition</i>[?Yield]
    <b>static</b> <i>MethodDefinition</i>[?Yield]
    <b>;</b>
</pre>

#### References:
<pre>
<i>Declaration</i>[Yield]:
    <i>ClassDeclaration</i><sub>[?Yield]</sub>

<i>ExportDeclaration</i> :
    <b>export</b> <b>default</b> <i>ClassDeclaration</i><sub>[Default]</sub>

<i>PrimaryExpression</i> :
    <i>ClassExpression</i>

</pre>

## Early Errors

```
ClassBody : ClassElementList
```

It is a Syntax Error if `PrototypePropertyNameList` of `ClassElementList` contains more than one occurrence of `"constructor"`.

```
ClassElement : MethodDefinition
```

It is a Syntax Error if `PropName` of `MethodDefinition` is not "constructor" and `HasDirectSuper(MethodDefinition)` is `true`.

It is a Syntax Error if `PropName` of `MethodDefinition` is "constructor" and `SpecialMethod` of `MethodDefinition` is `true`.

```
ClassElement : static MethodDefinition
```
It is a Syntax Error if `HasDirectSuper(MethodDefinition)` is true.
It is a Syntax Error if `PropName` of `MethodDefinition` is `"prototype"`.

### PR DESCRIPTION
**Pull request by ikarienator**
fixes #82.

### CODE REVIEW

**Comment by michaelficarra** [comment]
FYI, I've rebased off `es6`, which was recently rebased off `master` to get the location info update.

**Comment by ikarienator** [comment]
Added `locationSanityTest` for classes.

File: src/parser.js [hunk 1]
```diff
@@ -1551,55 +1548,55 @@ export class Parser extends Tokenizer {
   }
 
 
-  parseObjectExpressionItems(propertyMap) {
+  parseObjectExpressionItems() {
     let result = [];
+    let has__proto__ = [false];
```
Comment (michaelficarra) [inline src/parser.js]: Why is this a list?

File: src/parser.js [hunk 2]
```diff
@@ -1633,66 +1632,158 @@ export class Parser extends Tokenizer {
...
+  /**
+   * Try to parse a method definition.
+   *
+   * If it turns out to be one of:
+   *  * `IdentifierReference`
+   *  * `CoverInitializedName` (`IdentifierReference "=" AssignmentExpression`)
+   *  * `PropertyName : AssignmentExpression`
+   * The the parser will stop at the end of the leading `Identifier` or `PropertyName` and return it.
+   *
+   * @returns {{methodOrKey: (Shift.Method|Shift.PropertyName), kind: string}}
```
Comment (michaelficarra) [inline src/parser.js]: I'd rather see `{method: Shift.Method, kind: "method"} | {key: Shift.PropertyName, kind: "identifier"}`.

**Comment by michaelficarra** [comment]
Merging anyway. Who cares. What does life even mean?

**Pull request merged**